syntax = "proto3";

package apc;

message BaseMessage {
  string batch_id = 1;
  string sender_id = 2;
  int64 timestamp = 3;
}

message ProposeTask {
  BaseMessage base = 1;
  string step_name = 2;
  map<string, string> params = 3;
  string role = 4; // always "conductor"
}

message Accept {
  BaseMessage base = 1;
  string step_name = 2;
}

message Reject {
  BaseMessage base = 1;
  string step_name = 2;
  string reason = 3;
}

message Completed {
  BaseMessage base = 1;
  string step_name = 2;
  bool success = 3;
  map<string, string> result = 4;
}

message Failed {
  BaseMessage base = 1;
  string step_name = 2;
  string error_code = 3;
  string error_msg = 4;
}

message TakeOver {
  BaseMessage base = 1;
  string new_conductor_id = 2;
}
